cmd/clusterctl/cmd/generate_cluster.go

```diff
@@ -0,0 +1,188 @@
+		"Path to a kubeconfig file to use for the management cluster. If empty, default discovery rules apply.")
+	generateClusterClusterCmd.Flags().StringVar(&gc.kubeconfigContext, "kubeconfig-context", "",
+		"Context to be used within the kubeconfig file. If empty, current context will be used.")
+
+	// flags for the template variables
+	generateClusterClusterCmd.Flags().StringVarP(&gc.targetNamespace, "target-namespace", "n", "",
+		"The namespace to use for the workload cluster. If unspecified, the current namespace will be used.")
+	generateClusterClusterCmd.Flags().StringVar(&gc.kubernetesVersion, "kubernetes-version", "",
+		"The Kubernetes version to use for the workload cluster. If unspecified, the value from OS environment variables or the .cluster-api/clusterctl.yaml config file will be used.")
+	generateClusterClusterCmd.Flags().Int64Var(&gc.controlPlaneMachineCount, "control-plane-machine-count", 1,
+		"The number of control plane machines for the workload cluster.")
+	generateClusterClusterCmd.Flags().Int64Var(&gc.workerMachineCount, "worker-machine-count", 0,
+		"The number of worker machines for the workload cluster.")
+
+	// flags for the repository source
+	generateClusterClusterCmd.Flags().StringVarP(&gc.infrastructureProvider, "infrastructure", "i", "",
+		"The infrastructure provider to read the workload cluster template from. If unspecified, the default infrastructure provider will be used.")
+	generateClusterClusterCmd.Flags().StringVarP(&gc.flavor, "flavor", "f", "",
+		"The workload cluster template variant to be used when reading from the infrastructure provider repository. If unspecified, the default cluster template will be used.")
+
+	// flags for the url source
+	generateClusterClusterCmd.Flags().StringVar(&gc.url, "from", "",
+		"The URL to read the workload cluster template from. If unspecified, the infrastructure provider repository URL will be used")
+
+	// flags for the config map source
+	generateClusterClusterCmd.Flags().StringVar(&gc.configMapName, "from-config-map", "",
+		"The ConfigMap to read the workload cluster template from. This can be used as alternative to read from the provider repository or from an URL")
+	generateClusterClusterCmd.Flags().StringVar(&gc.configMapNamespace, "from-config-map-namespace", "",
+		"The namespace where the ConfigMap exists. If unspecified, the current namespace will be used")
+	generateClusterClusterCmd.Flags().StringVar(&gc.configMapDataKey, "from-config-map-key", "",
+		fmt.Sprintf("The ConfigMap.Data key where the workload cluster template is hosted. If unspecified, %q will be used", client.DefaultCustomTemplateConfigMapKey))
+
+	// other flags
+	generateClusterClusterCmd.Flags().BoolVar(&gc.listVariables, "list-variables", false,
+		"Returns the list of variables expected by the template instead of the template yaml")
+
+	generateCmd.AddCommand(generateClusterClusterCmd)
+}
+
+func runGenerateClusterTemplate(cmd *cobra.Command, name string) error {
```

Comment:
all of this seems to be duplicate code from cmd/clusterctl/cmd/config_cluster.go, can we instead reuse the same helper functions from generate_cluster into config_cluster.go so that the code is shared between the two commands?

Reply:
Since we are going anyway to remove `config_cluster.go` in the next release, I thought it'd be okay to just copy over. Same for the `config_provider.go` as well.

Reply:
If those are duplicates, let's start removing to existing methods in config_cluster.go/config_provider.go and have both the commands to point to the same internal function

Reply:
+1 @shysank are you ok with the proposed changes above?

Reply:
One clarification on the above comment. The --raw is for generate_provider. But the idea is still the same given that we are planning to remove config_cluster in 0.5.x, and there may be new options added to generate_cluster.

Reply:
given the context above (thanks for the explanation!), I'm fine with current implementation

Reply:
Same here, thanks for clarifying. We'll have to be diligent in reviews to make sure no changes are made to the config functions once this merges. It might be good to add a comment as well as Vince suggested.

Reply:
Yeah, I'll add comments. Thanks for the feedback!